keps/0019-package-api-versioning.md

```diff
@@ -0,0 +1,107 @@
+---
+kep-number: 19
+title: Versioning of Operator Packages
+authors:
+  - "@nfnt"
+owners:
+  - "@nfnt"
+creation-date: 2019-11-05
+last-updated: 2019-11-05
+status: provisional
+---
+
+# Versioning of Operator Packages
+
+## Table of Contents
+
+* [Summary](#summary)
+* [Motivation](#motivation)
+  * [Goals](#goals)
+  * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+  * [Support multiple API versions in the package repository](#support-multiple-api-versions-in-the-package-repository)
+  * [Install a specific application version](#install-a-specific-application-version)
+  * [Naming of package tarballs](#naming-of-package-tarballs)
+  * [Semantic versioning](#semantic-versioning)
+  * [Risks and Mitigations](#risks-and-mitigations)
+* [Implementation History](#implementation-history)
+
+## Summary
+
+The different versions of components that describe an operator package are surfaced to users, so that they are able to install exactly specify which version can be installed. Furthermore, compatibility with existing packages for older APIs of KUDO is provided.
+
+## Motivation
+
+The dependencies between KUDO, operators packages and the application that is managed by the operator enforce the use of multiple versions to fully describe these dependencies. First, there is the dependency on a specific API version of KUDO. Also, a specific version of an application is provided by the package. And last, the operator itself may have multiple revisions. All of these versions need to be captured as part of the operator in a concise way that makes it easy for users to install specific operators for any of these versions.
+
+### Goals
+
+* Allow users to install packages from a package repository that provides packages for multiple KUDO API versions
+* Allow users to install a specific version of an application and/or a specific revision of an operator
+* Reject packages that don't have a matching KUDO API version
+
+### Non-Goals
+
+* Handle operator dependencies on specific Kubernetes versions
+
+## Proposal
+
+A `operator.yaml` file already provides versions for all components describing an operator:
+  * The dependency on KUDO is provided by the `apiVersion` and `kudoVersion` fields
+  * The bundled application version is provided by the `appVersion` field
+  * The version of the operator is provided by the `version` field
+  * The version of Kubernetes is provided by the `kubernetesVersion` field
+
+From these versions, currently only the `version` field can be used when installing packages with `kubectl kudo install`. Furthermore, a package repository index only contains `appVersion` and `version` fields in its entries. This limits the possible installation scenarios.
+At least the following installation scenarios should be supported:
+  1. A package repository could contain operators for older `apiVersion` and than the currently released KUDO. A user should be able to install these operators with older versions of KUDO.
```

Comment:
I agree with this statement in general.. however :) .   it isn't the whole story.    A version of KUDO should have awareness of which `apiVersions`  it supports.   Then I expect that it could at some point deprecate an `apiVersion`.   It would be better to state the problem this way IMO.   kudo 2.1.1 supports apiVersions 2.0, 2.1, 2.2 AND support 1.9 in deprecated warning mode AND does NOT support <= 1.8.   In this way kudo should have 3 apiVersions sets it manages... 

1.  active support
2. deprecate support
3. no longer supported

Reply:
Good point! I'll clarify this in the KEP.

Reply:
After thinking about it for 18 days 😆, IMO this shouldn't be part of this KEP. Let's focus this KEP on operator packages. How and which API versions are supported by KUDO and what a deprecation cycle for these APIs would look like should be part of a separate KEP. @kensipe wdyt?